pytorch_lightning/core/memory.py

```diff
@@ -1,164 +1,243 @@
+    @property
+    def named_modules(self) -> List[Tuple[str, nn.Module]]:
+        if self._mode == ModelSummary.MODE_FULL:
+            mods = self._model.named_modules()
             mods = list(mods)[1:]  # do not include root module (LightningModule)
-        elif self.mode == 'top':
+        elif self._mode == ModelSummary.MODE_TOP:
             # the children are the top-level modules
-            mods = self.model.named_children()
+            mods = self._model.named_children()
         else:
             mods = []
         return list(mods)
 
-    def get_variable_sizes(self) -> None:
-        """ Run sample input through each layer to get output sizes. """
-        mods = self.named_modules()
-        in_sizes = []
-        out_sizes = []
-        input_ = self.model.example_input_array
+    @property
+    def layer_names(self) -> List[str]:
+        return list(self._layer_summary.keys())
 
-        if self.model.on_gpu:
-            device = next(self.model.parameters()).get_device()
-            # test if input is a list or a tuple
-            if isinstance(input_, (list, tuple)):
-                input_ = [input_i.cuda(device) if torch.is_tensor(input_i) else input_i
-                          for input_i in input_]
-            else:
-                input_ = input_.cuda(device)
+    @property
+    def layer_types(self) -> List[str]:
+        return [layer.layer_type for layer in self._layer_summary.values()]
 
-        if self.model.trainer.use_amp:
-            # test if it is not a list or a tuple
-            if isinstance(input_, (list, tuple)):
-                input_ = [input_i.half() if torch.is_tensor(input_i) else input_i
-                          for input_i in input_]
-            else:
-                input_ = input_.half()
+    @property
+    def in_sizes(self) -> List:
+        return [layer.in_size for layer in self._layer_summary.values()]
+
+    @property
+    def out_sizes(self) -> List:
+        return [layer.out_size for layer in self._layer_summary.values()]
+
+    @property
+    def param_nums(self) -> List[int]:
+        return [layer.num_parameters for layer in self._layer_summary.values()]
 
+    def summarize(self) -> Dict[str, LayerSummary]:
+        summary = OrderedDict((name, LayerSummary(module)) for name, module in self.named_modules)
+        if self._model.example_input_array is not None:
+            self._forward_example_input()
+        return summary
+
+    def _forward_example_input(self) -> None:
+        """ Run the example input through each layer to get input- and output sizes. """
+        model = self._model
+        trainer = self._model.trainer
+
+        input_ = model.example_input_array
+        input_ = model.transfer_batch_to_device(input_, model.device)
+        input_ = apply_to_collection(input_, torch.Tensor, lambda x: x.type(model.dtype))
+
+        if trainer is not None and trainer.use_amp:
+            if model.use_native_amp:
```

Comment:
`use_native_amp==True` does not imply that `use_amp==True`, they are independent, I agree that it shall be as you write...